File: cpp/permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        dfs(nums, 0);
        return res;
    }
    void dfs(vector<int> &nums, int index) {
        if (index == nums.size()) {
            res.push_back(nums);
            return;
        }
        for(int i = index; i < nums.size(); i++) {
            swap(nums[index], nums[i]);
            dfs(nums, index+1);
            swap(nums[index], nums[i]);
        }
    }
private:
    vector<vector<int>> res;
};

File: python/permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(0, nums, res)
        return res
        
    def dfs(self, i, nums, res):
        if i >= len(nums):
            res.append(nums[:])
            return
        for j in xrange(i, len(nums)):
            nums[i], nums[j] = nums[j], nums[i]
            self.dfs(i+1, nums, res)
            nums[i], nums[j] = nums[j], nums[i]
